components/camel-pg-replication-slot/src/main/java/org/apache/camel/component/pg_replication_slot/PgReplicationSlotConsumer.java

```diff
@@ -0,0 +1,257 @@
+
+            this.payload = null;
+        } catch (PSQLException e) {
+            validateConnection(e);
+        } catch (SQLException e) {
+            getExceptionHandler().handleException("Exception while sending feedback to PostgreSQL.", exchange, e);
+        }
+    }
+
+    private void processRollback(Exchange exchange) {
+        Exception cause = exchange.getException();
+        if (cause != null) {
+            getExceptionHandler().handleException("Error during processing exchange. Will attempt to process the message on next poll.", exchange, cause);
+        }
+    }
+
+    private void createSlot() throws SQLException {
+        this.pgConnection.getReplicationAPI()
+                .createReplicationSlot()
+                .logical()
+                .withSlotName(this.endpoint.getSlot())
+                .withOutputPlugin(this.endpoint.getOutputPlugin())
+                .make();
+    }
+
+    private boolean isSlotCreated() throws SQLException {
+        String sql = String.format("SELECT count(*) FROM pg_replication_slots WHERE slot_name = '%s';", this.endpoint.getSlot());
+
+        try (Statement statement = this.connection.createStatement(); ResultSet resultSet = statement.executeQuery(sql)) {
+            resultSet.next();
+            return resultSet.getInt(1) > 0;
+        }
+    }
+
+    private PGReplicationStream getStream() throws SQLException {
+        if (this.replicationStream != null && !this.replicationStream.isClosed()) {
+            return this.replicationStream;
+        }
+
+        if (isSlotActive()) {
+            log.info(String.format("Slot: %s is active. Waiting for it to be available.", this.endpoint.getSlot()));
+            return null;
+        }
+
+        ChainedLogicalStreamBuilder streamBuilder = this.pgConnection.getReplicationAPI()
+                .replicationStream()
+                .logical()
+                .withSlotName(this.endpoint.getSlot())
+                .withStatusInterval(this.endpoint.getStatusInterval(), TimeUnit.SECONDS);
+
+        Properties slotOptions = new Properties();
+        slotOptions.putAll(this.endpoint.getSlotOptions());
+        streamBuilder.withSlotOptions(slotOptions);
+
+        this.replicationStream = streamBuilder.start();
+
+        return this.replicationStream;
+    }
+
+    private boolean isSlotActive() throws SQLException {
+        String sql = String.format("SELECT count(*) FROM pg_replication_slots where slot_name = '%s' AND active = true;",
+                this.endpoint.getSlot());
+
+        try (Statement statement = this.connection.createStatement(); ResultSet resultSet = statement.executeQuery(sql)) {
+            resultSet.next();
+            return resultSet.getInt(1) > 0;
+        }
+    }
+
+    private void connect() throws SQLException {
+        if (this.connection != null) {
+            this.connection.close();
+        }
+
+        this.connection = this.endpoint.newDbConnection();
```

Comment:
will there be a single connection, or a pool of connections can be considered?

Reply:
Replication stream connections by nature are long-running single-threaded (to preserve order) connections. So, I don't think using a connection pool will benefit the application as the streams won't be released frequently. We use the approach above in production for a long time and we didn't face any problems so far. I'll give it more research time though.

Reply:
fair enough. that may be an improvement which can be done later